Source/Scene/parseBatchTable.js

```diff
@@ -0,0 +1,275 @@
+import MetadataSchema from "./MetadataSchema.js";
+
+/**
+ * An object that parses the the 3D Tiles 1.0 batch table and transcodes it to
+ * be compatible with feature metadata from the `EXT_feature_metadata` glTF extension
+ * <p>
+ * See the {@link https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_feature_metadata/1.0.0|EXT_feature_metadata Extension} for glTF.
+ * </p>
+ *
+ * @param {Object} options Object with the following properties:
+ * @param {Number} options.count The number of features in the batch table.
+ * @param {Object} options.batchTable The batch table JSON
+ * @param {Uint8Array} [options.binaryBody] The batch table binary body
+ * @return {FeatureMetadata} a transcoded feature metadata object
+ *
+ * @private
+ */
+export default function parseBatchTable(options) {
+  //>>includeStart('debug', pragmas.debug);
+  Check.typeOf.number("options.count", options.count);
+  Check.typeOf.object("options.batchTable", options.batchTable);
+  //>>includeEnd('debug');
+
+  var featureCount = options.count;
+  var batchTable = options.batchTable;
+  var binaryBody = options.binaryBody;
+
+  // divide properties into binary, json and hierarchy
+  var partitionResults = partitionProperties(batchTable);
+
+  var jsonMetadataTable = new JsonMetadataTable({
+    count: featureCount,
+    properties: partitionResults.jsonProperties,
+  });
+
+  var hierarchy = initializeHierarchy(partitionResults.extensions, binaryBody);
+
+  var binaryResults = transcodeBinaryProperties(
+    featureCount,
+    partitionResults.binaryProperties,
+    binaryBody
+  );
+
+  var featureTables = {
+    _batchTable: new FeatureTable({
+      featureTable: binaryResults.featureTableJson,
+      class: binaryResults.transcodedClass,
+      bufferViews: binaryResults.bufferViewsU8,
+      jsonMetadataTable: jsonMetadataTable,
+      batchTableHierarchy: hierarchy,
+    }),
+  };
+
+  return new FeatureMetadata({
+    schema: binaryResults.transcodedSchema,
+    featureTables: featureTables,
+    extensions: partitionResults.extensions,
+    extras: partitionResults.extras,
+  });
+}
+
+/**
+ * Divide the batch table's properties into binary, JSON and hierarchy
+ * extension as each is handled separately
+ *
+ * @param {Object} batchTable The batch table JSON
+ * @returns {Object} The batch table divided into binary, JSON and hierarchy portions.
+ *
+ * @private
+ */
+function partitionProperties(batchTable) {
+  var legacyHierarchy = batchTable.HIERARCHY;
+  var extras = batchTable.extras;
+  var extensions = batchTable.extensions;
+
+  if (defined(legacyHierarchy)) {
+    parseBatchTable._deprecationWarning(
+      "batchTableHierarchyExtension",
+      "The batch table HIERARCHY property has been moved to an extension. Use extensions.3DTILES_batch_table_hierarchy instead."
+    );
+    extensions = defined(extensions) ? extensions : {};
+    extensions["3DTILES_batch_table_hierarchy"] = legacyHierarchy;
```

Comment:
It might be cleaner if `partionResults` holds onto the hierarchy object directly instead of modifying `extensions`